autoscaler/autoscaler_builder.go

```diff
@@ -0,0 +1,115 @@
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util/clock"
+)
+
+// Scaler determines the number of replicas to run
+type AutoScaler struct {
+	Verbose      bool
+	Clock        clock.Clock
+	ConfigMapKey string
+	PollPeriod   time.Duration
+	K8sClient    K8sClientInterface
+	StopCh       chan struct{}
+	ReadyCh      chan<- struct{} // For testing.
+
+	version       string
+	coreLadder    []int
+	coreScalerMap map[int]int
+	nodeLadder    []int
+	nodeScalerMap map[int]int
+}
+
+func (s *AutoScaler) calculateExpectedNumOfReplicas(clusterStatus ClusterStatus, configMap ConfigMap) (int32, error) {
+
+	params, err := parseScalerParams([]byte(configMap.data))
+	if err != nil {
+		return 0, fmt.Errorf("error parsing scaler params: %s", err)
+	}
+
+	// Get the expected replicas for the currently schedulable nodes and cores
+	expReplicas := int32(s.scaleWithNodesAndCores(params, configMap.version, int(clusterStatus.schedulableNodes), int(clusterStatus.schedulableCores)))
+
+	return expReplicas, nil
+}
+
+func (s *AutoScaler) buildScalerMap(resourceType string, params []scalerEntry) (map[int]int, []int) {
+
+	scalerMap := make(map[int]int)
+	ladder := make([]int, 0, len(params))
+	// construct a search coreLadder from the map
+	for _, e := range params {
+		k := e[0] // the first element should be the actual number of resources
+		v := e[1] // the second element should be the desired number of replicas
+		scalerMap[k] = v
+		ladder = append(ladder, k)
+		log.Printf("%v > %5d => %3d replicas\n", resourceType, k, v)
+	}
+	sort.Ints(ladder)
+	return scalerMap, ladder
+}
+
+func (s *AutoScaler) scaleWithNodesAndCores(params *scalerParams, resourceVersion string, numCurrentNodes, schedulableCores int) int {
+
+	if !(resourceVersion == s.version) {
+		// Only rebuild our tables if the configmap changed
+		log.Printf("Detected ConfigMap version change (old: %s new: %s) - rebuilding lookup tables\n", s.version, resourceVersion)
+		s.coreScalerMap, s.coreLadder = s.buildScalerMap("cores", params.CoresToReplicasMap)
+		s.nodeScalerMap, s.nodeLadder = s.buildScalerMap("nodes", params.NodesToReplicasMap)
+		s.version = resourceVersion
+	}
+	return s.scalerLookup(numCurrentNodes, schedulableCores)
+}
+
+func (s *AutoScaler) scalerLookup(schedulableNodes, schedulableCores int) int {
+
+	replicasFromNodeMap := getExpectedReplicasFromMap(schedulableNodes, s.nodeLadder, s.nodeScalerMap)
+	replicasFromCoreMap := getExpectedReplicasFromMap(schedulableCores, s.coreLadder, s.coreScalerMap)
+
+	// Returns the lookup which yields the most replicas
+	return getMaxInt(replicasFromCoreMap, replicasFromNodeMap)
+}
+
+func getExpectedReplicasFromMap(schedulableResources int, ladder []int, scalerMap map[int]int) int {
+
+	neededReplicas := 1
+	// binary search for the corresponding replicas number
+	pos := sort.SearchInts(ladder, schedulableResources)
+	if pos == len(ladder) {
```

Comment:
Do we need all of these cases? It seems like all you need to check for is if the pos is past the end of the array, then use the maximum value.

Probably also want to emit a warning if the maximum value is used...